Frontend settings support for configuring "playground" feature

sumanthvrao opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This picks up where #16496 finished and adds frontend support for configuring playgrounds. The third, fourth, and fifth commits add the majority of the UI for reading, creating and deleting playgrounds. The issue filed for the feature is #16432.

**Testing plan:** <!-- How have you tested? -->
Tested manually and added puppeteer frontend tests for the playground UI.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![image](https://user-images.githubusercontent.com/30312566/115808861-b64b1680-a408-11eb-9b2e-bd097d338e24.png)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->
Here is a list of pending TODO items (ordered on priority):
- [x] We need to implement a typeahead popup for the pygments_language field in the playgrounds form. (I'll probably do this soon)
- [ ] Add a way for organizations that want to disable the feature (due to concerns of sending code to a third party) and document it.
- [ ] Investigate what services like `repl.it` exist and contact a few for permission when thinking about what defaults to support. 
- [ ] Also, now would be a good time to discuss how we want to manage the default playgrounds for an organization.
- [ ] Support editing the playground links (similar to #17045) - we may want to push this off for a bit longer.

sumanthvrao commented:
@timabbott This is ready for a round of review :)

timabbott commented on static/templates/admin_playground_list.hbs:
```diff
@@ -9,5 +9,12 @@
+    {{#if ../can_modify}}
+    <td class="no-select actions">
+        <button class="button small delete btn-danger" data-playground-id="{{id}}">
+            <i class="fa fa-trash-o" aria-hidden="true"></i>
```
This should be `aria-label` + `title` -- you want `aria-hidden` for icons that are decorations to a textual label for a button; and aria-label+title when the icon is the entire button.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA has background for this.

timabbott commented on static/js/settings_playgrounds.js:
```diff
@@ -88,6 +90,23 @@ function build_page() {
+            success() {
+                const row = btn.parents("tr");
+                row.remove();
+            },
```
This success function should do nothing -- the row should be removed by the `events` handler.

timabbott commented on static/templates/admin_playground_list.hbs:
```diff
@@ -0,0 +1,13 @@
+        <span>{{url_prefix}}</span>
+    </td>
+</tr>
+{{/with}}
```
This probably belongs in `static/templates/settings/`.

timabbott replied:
(Though it looks like we have like 10 we should move).

timabbott replied:
Cool, and opened https://github.com/zulip/zulip/issues/18227 for the rest.

timabbott commented on static/templates/settings/playground_settings_admin.hbs:
```diff
@@ -0,0 +1,21 @@
+        <p>
+            {{#tr}}
+            Configure playground options to open code appearing in message code blocks
+            in these external playgrounds.
```
Perhaps: 
"Configure external sites that can be used to preview or run content present in a code block."

timabbott commented on templates/zerver/app/settings_overlay.html:
```diff
@@ -134,6 +134,13 @@ <h1>{{ _('Settings') }}<span class="section"></span></h1>
                 </li>
+                <li tabindex="0" data-section="playground-settings">
+                    <i class="icon fa fa-external-link" aria-hidden="true"></i>
+                    <div class="text">{{ _('Playgrounds') }}</div>
```
Let's call this "Code playgrounds" in the menu -- that'll provide important context for what this is for a user who has no idea what a "playground" is.

timabbott commented on static/templates/settings/playground_settings_admin.hbs:
```diff
@@ -6,6 +6,34 @@
+                    <div class="control-group">
+                        <label for="playground_name" class="control-label"> {{t "Name" }} </label>
+                        <input type="text" id="playground_name" name="name" placeholder="Python3 playground" />
+                    </div>
```
If it's convenient, we may want to add "hint" lines to this form explaining what these parameters are.

sumanthvrao replied:
Do you mean hints like these?

![image](https://user-images.githubusercontent.com/30312566/115334783-9d98f180-a1b9-11eb-8339-d236940a2609.png)

timabbott replied:
Yeah (I'm not sure I love the visual style, but it's serviceable and we can always iterate)

timabbott commented on templates/zerver/features.html:
```diff
@@ -64,6 +64,13 @@ <h3>CUSTOMIZABLE AUTOMATIC LINKIFICATION</h3>
+                <h3>EXTERNAL CODE PLAYGROUND OPTIONS</h3>
+                <p>
+                    Configure playground options to open code in a Zulip
+                    message's code block using external playgrounds.
```
This description is probably going to be too opaque for a random user.  It's possible we just want to this advertised on /for/open-source and on the "code blocks" section of the markdown advice, since it's a little hard to explain and I'm not sure how much non-OSS organizations will want this.

@alya thoughts?

alya replied:
Hmm, good question... Are you suggesting not having a way to set it up in the menus at all?

In general, I think we should make sure we never use the term "playground" by itself, as it would be unfamiliar and confusing to many users. "Code playground" at least makes it clear for non-coders that "it's a thing for code that I don't care about".

I suppose we could also make the menu option more generically say "Code blocks". I think it would be less weird for users who are unfamiliar with playgrounds, though more opaque for users who are.

timabbott replied:
No, I was suggesting we not try to explain this feature on /features (aka that specific landing page).

I think a "Code blocks" title here with the text explaining the "Copy code block content" button and external playgrounds would be reasonable.

sumanthvrao replied:
Cool! Updated the /features page with a "Code blocks" section.

timabbott commented on templates/zerver/for/open-source.md:
```diff
@@ -150,6 +150,11 @@ Efficiently refer to issues or code reviews with notation like `#1234` or
+### Open code in external playgrounds.
+
+Configure [external code playgrounds](/help/add-a-custom-playground) in which
+you can easily open and run code occuring in a Zulip message's code block.
```
I think my best idea for how to explain this feature is this:

"Instantly transfer content from a [Zulip code block](/help/...) to an [external code playgrounds](/help/...) that can run the code."

alya replied:
Typo in Tim's comment -- it should be "playground", not "playgrounds".

+1 for making sure that a user who's never heard of playgrounds can learn what they are by reading the instructions.

sumanthvrao replied:
Sounds great! Also, do we want to stick the same positioning (as in this commit) of where the feature will appear in the list of features in /for/open-source?

timabbott commented on templates/zerver/help/add-a-custom-playground.md:
```diff
@@ -0,0 +1,46 @@
+~~~
+
+The user would then get a on-hover option to open the above code in the playground
+they had previously configured.
```
We should add a "Technical details" section that mentions:
* You can configure multiple playgrounds for a given language; if you do that, the user will get to choose which playground to open the code in using the name.
* The "Language" field is the human-readable Pygments language for the programming language, and some brief description of how the mapping works (or how to see them).
* That you can make up your own language and configure a playground for that, for something custom.

timabbott commented:
I ended up merging one of the settings commits after resolving my feedback and adding an `if development_environment` block to the templates to avoid the settings page appearing in production before the feature is fully implemented.

sumanthvrao commented on static/js/settings_playgrounds.js:
```diff
@@ -84,4 +87,35 @@ export function set_up() {
+                        $t_html({defaultMessage: "Custom playground added!"}),
+                        playground_status,
+                    );
+                    // TODO: Should we do a reload here for the codeblocks to re-render?
```
Would it be okay if we take this up in v2 of playgrounds? If so, I'll leave a more elaborate `TODO` in the code with your thoughts on the few directions we can proceed with this.

timabbott replied:
Yes, definitely fine to leave this as a follow-up issue.

sumanthvrao commented on static/templates/settings/playground_settings_admin.hbs:
```diff
@@ -6,6 +6,11 @@
         </p>
+        <p>
+            {{#tr}}
+            More details are available <a href="/help/add-a-custom-playground" target="_blank" rel="noopener noreferrer">in the Help Center article</a>.
```
`help_link_widget` seems to fit nicely after the "Add a new playground" form header.
![image](https://user-images.githubusercontent.com/30312566/115328731-ba7bf780-a1ae-11eb-96dd-426ab38d52d5.png)

sumanthvrao commented:
@timabbott Thanks for the review and merging parts of this! I left some replies to the inline comments you left and have pushed an update addressing the points you mentioned.

timabbott commented:
@sumanthvrao it looks like the puppeteer tests for this are failing?

timabbott commented:
I think aside from the CI issue, and I guess a rebase for the OpenAPI docs, this just needs another pass on documentation polish, which perhaps @alya can help with tomorrow.

sumanthvrao commented:
@timabbott Thanks for the follow-up. The last commit here adds the Typeahead UI in the frontend. Looks like there was a race condition that caused the puppeteer tests in the CI to fail. The below patch fixes it. I also added a commit at the end to remove the `development_environment` toggle in the settings UI.
```
---
 frontend_tests/puppeteer_tests/realm-playground.ts | 7 ++++++-
 static/js/settings_playgrounds.js                  | 8 +++++++-
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/frontend_tests/puppeteer_tests/realm-playground.ts b/frontend_tests/puppeteer_tests/realm-playground.ts
index 875fd04c5..61e0dc583 100644
--- a/frontend_tests/puppeteer_tests/realm-playground.ts
+++ b/frontend_tests/puppeteer_tests/realm-playground.ts
@@ -12,10 +12,15 @@ type Playground = {
 
 async function _add_playground_and_return_status(page: Page, payload: Playground): Promise<string> {
     await page.waitForSelector(".admin-playground-form", {visible: true});
+    // Let's first ensure that the success/failure status from an earlier step has disappeared.
+    const admin_playground_status_selector = "div#admin-playground-status";
+    await page.waitForSelector(admin_playground_status_selector, {hidden: true, timeout: 4000});
+
+    // Now we can fill and click the submit button.
     await common.fill_form(page, "form.admin-playground-form", payload);
     await page.click("form.admin-playground-form button.button");
 
-    const admin_playground_status_selector = "div#admin-playground-status";
+    // We return the success/failure status message back to the caller.
     await page.waitForSelector(admin_playground_status_selector, {visible: true});
     const admin_playground_status = await common.get_text_from_selector(
         page,
diff --git a/static/js/settings_playgrounds.js b/static/js/settings_playgrounds.js
index 31f4f6730..e1ccf90a0 100644
--- a/static/js/settings_playgrounds.js
+++ b/static/js/settings_playgrounds.js
@@ -128,6 +128,7 @@ function build_page() {
                     ui_report.success(
                         $t_html({defaultMessage: "Custom playground added!"}),
                         playground_status,
+                        3000,
                     );
                     // FIXME: One thing to note here is that the "view code in playground"
                     // option for an already rendered code block (tagged with this newly added
@@ -141,7 +142,12 @@ function build_page() {
                 },
                 error(xhr) {
                     add_playground_button.prop("disabled", false);
-                    ui_report.error($t_html({defaultMessage: "Failed"}), xhr, playground_status);
+                    ui_report.error(
+                        $t_html({defaultMessage: "Failed"}),
+                        xhr,
+                        playground_status,
+                        3000,
+                    );
                 },
             });
         });
-- 
2.17.1
```

sumanthvrao commented:
Just re-ordered the commits and changed the commit description for a couple of them.

sumanthvrao commented:
Oh no, hmmmmm.... I'm not sure why the puppeteer test is failing again due to a timeout issue. I might be missing out on something here. @timabbott any ideas?
```
async function _add_playground_and_return_status(page: Page, payload: Playground): Promise<string> {
    await page.waitForSelector(".admin-playground-form", {visible: true});
    // Let's first ensure that the success/failure status from an earlier step has disappeared.
    const admin_playground_status_selector = "div#admin-playground-status";
    await page.waitForSelector(admin_playground_status_selector, {hidden: true, timeout: 4000});

    // Now we can fill and click the submit button.
    await common.fill_form(page, "form.admin-playground-form", payload);
    await page.click("form.admin-playground-form button.button");

    // We return the success/failure status message back to the caller.
    await page.waitForSelector(admin_playground_status_selector, {visible: true});
    const admin_playground_status = await common.get_text_from_selector(
        page,
        admin_playground_status_selector,
    );
    return admin_playground_status;
}
```

alya commented on static/templates/settings/playground_settings_admin.hbs:
```diff
@@ -5,6 +5,43 @@
             Configure external sites that can run code present in a Zulip code block.
```
I would change this to: "Configure external sites that can run code from a Zulip code block."

sumanthvrao replied:
We do link to the Help Center article from the blue question-mark symbol on the page.

alya replied:
Ah, good point! I guess whether or not we want to mention the Help Center in the text may depend on whether it's helpful for understanding the overall context vs. just instructions for how to add a new playground (what I'd expect to see under `?`).

sumanthvrao commented:
I have extracted the puppeteer tests into a commit (the last one in the series). I also tweaked the test, where previously a `page.click` was used to click on the playground submit form button, but now happens using a `page.$eval` (which seems to be passing CI). 

However, we can choose to merge the rest of the commits for now if they are all OK.

alya reviewed (commented):
I went through the settings UI; will comment on the documentation page separately.

timabbott commented:
This lgtm aside from the documentation stuff, so I'm hoping to merge it to master in the next couple days.

sumanthvrao commented on static/templates/settings/playground_settings_admin.hbs:
```diff
@@ -5,6 +5,43 @@
+                        {{!-- TODO: This should have a typeahead suggestion popover --}}
+                        <label for="playground_pygments_language" class="control-label"> {{t "Language" }}
+                            <i class="fa fa-info-circle settings-info-icon tippy-zulip-tooltip"
+                              aria-hidden="true" data-tippy-placement="right" data-tippy-content="{{t 'The human-readable Pygments language name for the programming language' }}"></i>
```
- Yeah, if we indeed explain the playgrounds feature in more details (as you have in the previous comment), then I don't see a need for the `i` buttons.
- The typeahead for the code blocks is a list of pygments alias. In order for us to show playgrounds per programming language, we map these playgrounds into a user-friendly canonical name (E.G: py2 and python2 point to the "Python 2.x"). 
- I'll see if I can tweak the language selector typeahead based on your suggestions :)

sumanthvrao replied:
@alya I don't see any auto-complete in my setup. Is this a Zulip auto-complete or from some third-party software?

alya replied:
This is auto-complete form my browser. You can try to replicate by, e.g., creating a Python playground, and then going to create a second one and starting to type "p" for the language.

alya replied:
> The typeahead for the code blocks is a list of pygments alias. In order for us to show playgrounds per programming language, we map these aliases into a user-friendly canonical name (E.G: py2 and python2 point to the "Python 2.x"). Hence in both the backend and frontend, we can configure only one playground per friendly name.

Thinking more about it, I'm worried that the mismatch between the code blocks and the playground menu will be confusing. If what you type in the compose box doesn't match the playground menu, how do you know whether it will trigger the playground or not?

Maybe we can put all the pygment aliases into a parenthetical? E.g. "Python 2.x (py2, python2)"?

@timabbott  what do you think?

timabbott replied:
Yeah, I think that'd be a nice touch if we can do it.

timabbott replied:
(You can disable the browser autocomplete with `autocomplete="off"` -- we do that in many settings inputs for this reason.

alya commented:
Cool, thanks!

On Mon, May 3, 2021 at 6:05 PM Sumanth V Rao ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In static/templates/settings/playground_settings_admin.hbs
> <https://github.com/zulip/zulip/pull/18212#discussion_r625456826>:
>
> > @@ -5,6 +5,43 @@
>              Configure external sites that can run code present in a Zulip code block.
>              {{/tr}}
>          </p>
> +
> +        {{#if is_admin}}
> +        <form class="form-horizontal admin-playground-form">
> +            <div class="add-new-playground-box grey-box">
> +                <div class="new-playground-form wrapper">
> +                    <div class="settings-section-title">{{t "Add a new playground" }}</div>
> +                    <div class="alert" id="admin-playground-status"></div>
> +                    <div class="control-group">
> +                        {{!-- TODO: This should have a typeahead suggestion popover --}}
> +                        <label for="playground_pygments_language" class="control-label"> {{t "Language" }}
> +                            <i class="fa fa-info-circle settings-info-icon tippy-zulip-tooltip"
> +                              aria-hidden="true" data-tippy-placement="right" data-tippy-content="{{t 'The human-readable Pygments language name for the programming language' }}"></i>
>
>
>    - Yeah, if we indeed explain the playgrounds feature in more details
>    (as you have in the previous comment), then I don't see a need for the
>    i buttons.
>    - The typeahead for the code blocks is a list of pygments alias. In
>    order for us to show playgrounds per programming language, we map these
>    playgrounds into a user-friendly canonical name (E.G: py2 and python2 point
>    to the "Python 2.x").
>    - I'll see if I can tweak the language selector typeahead based on
>    your suggestions :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/18212#discussion_r625456826>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP6IUV4N4525J3JRDR36E3TL5B5VANCNFSM43F5AWCA>
> .
>

sumanthvrao commented on static/templates/settings/playground_settings_admin.hbs:
```diff
@@ -5,6 +5,43 @@
+                    <div class="control-group">
+                        <label for="playground_url_prefix" class="control-label"> {{t "URL prefix" }}
+                            <i class="fa fa-info-circle settings-info-icon tippy-zulip-tooltip"
+                              aria-hidden="true" data-tippy-placement="right" data-tippy-content="{{t 'The playground url prefix along with any query paramaters needed' }}"></i>
```
Great idea! I added a couple of examples for languages like Python and JavaScript.

sumanthvrao commented:
Pushed back to this with the documentation changes suggested by @alya (Thanks for the review!). Here's a screenshot of how the settings page looks with the new text introduction, examples and after removing the `i` icons.
![image](https://user-images.githubusercontent.com/30312566/117013365-6c6a0680-ad0d-11eb-9204-f84d559ab3ec.png)-+

alya commented:
Screenshot looks good, thanks! I think we can merge and have the rest of the comments be follow-ups.

sumanthvrao commented:
Just pushed back with the `autocomplete=off` change. This would leave us with the following as followups:

- Put the pygment aliases into a parenthetical E.g. "Python 2.x (py2, python2)"
- Add typeahead suggesting 5 popular languages before the user starts typing.
- A new `Code blocks` page under `Sending Messages` (being worked on by Alya)
- Add the remaining stuff mentioned in #16432

timabbott commented:
OK I've merged this PR!  I did a bit of work on essentially all the documentation before merging:

```
diff --git a/templates/zerver/features.html b/templates/zerver/features.html
index 53e25d43c0..40c28b8efd 100644
--- a/templates/zerver/features.html
+++ b/templates/zerver/features.html
@@ -67,11 +67,11 @@
             <div class="feature-block">
                 <h3>CODE BLOCKS</h3>
                 <p>
-                    With a single click one can easily copy code block
-                    contents to the clipboard. And transfer the contents
-                    to an external code playground that can run the code.
-                    Hover over a Zulip message's code block to see the
-                    options.
+                    With a single click, one can easily copy code block
+                    contents to the clipboard or transfer the contents
+                    to
+                    an <a href="/help/add-a-custom-playground">external
+                    code playground</a> that can run the code.
                 </p>
             </div>
         </div>
diff --git a/templates/zerver/for/open-source.md b/templates/zerver/for/open-source.md
index 077ceea34e..cfc7562213 100644
--- a/templates/zerver/for/open-source.md
+++ b/templates/zerver/for/open-source.md
@@ -126,11 +126,16 @@ Import your existing organization from [Slack](/help/import-from-slack),
 [Mattermost](/help/import-from-mattermost), or
 [Gitter](/help/import-from-gitter).
 
-### Syntax highlighting.
+### Collaborate on code and formulas
 
-[Full Markdown support](/help/format-your-message-using-markdown), including
-syntax highlighting, makes it easy to discuss code, paste an error message,
-or explain a complicated point. Full LaTeX support as well.
+[Markdown code blocks](/help/format-your-message-using-markdown#code)
+with syntax highlighting make it easy to discuss code, paste an error
+message, or explain a complicated point. Native LaTeX support provides
+the same benefits when talking about math.
+
+You can also instantly copy a code block to your clipboard or transfer
+it to an [external code playground](/help/add-a-custom-playground) to
+interactively run and debug the code.
 
 If your community primarily uses a single programming language,
 consider setting a default language for syntax highlighting.
@@ -150,11 +155,6 @@ Efficiently refer to issues or code reviews with notation like `#1234` or
 [custom linkification filter](/help/add-a-custom-linkifier) for
 your organization.
 
-### Open code in external playgrounds.
-
-Instantly transfer content from a [Zulip code block](/help/format-your-message-using-markdown#code)
-to an [external code playground](/help/add-a-custom-playground) that can run the code.
-
 ### Hundreds of integrations.
 
 Get events from GitHub, Travis CI, JIRA, and
diff --git a/templates/zerver/help/add-a-custom-playground.md b/templates/zerver/help/add-a-custom-playground.md
index 60c0b30820..eb60e72ed7 100644
--- a/templates/zerver/help/add-a-custom-playground.md
+++ b/templates/zerver/help/add-a-custom-playground.md
@@ -2,12 +2,18 @@
 
 {!admin-only.md!}
 
-Playgrounds allow users to open the contents of a message code block in
-an external playground of their choice. This decision is offered as a pop up
-when hovering over any Zulip message containing a code block. Playgrounds
-can be configured specific to a programming language and the same language
-can be specified using the [syntax highlighting format](https://zulip.com/help/format-your-message-using-markdown#code)
-when composing the message using Markdown.
+Code playgrounds allow users to open the contents of a [code
+block][code-block] in an external playground of your choice by
+clicking a widget that appears when hovering over the code block.
+
+Code playgrounds can be configured for any programming language (the
+language of a code block is determined by the logic for syntax
+highlighting).  You can also use a custom language name to implement
+simple integrations.  For example, a playground for the language
+`send_tweet` could be used with a "playground" that sends the content
+of the code block as a Tweet.
+
+[code-block]: /help/format-your-message-using-markdown#code
 
 ### Add a custom playground
 
@@ -48,17 +54,18 @@ they had previously configured.
 ## Technical details
 
 * You can configure multiple playgrounds for a given language; if you do that,
-  the user will get to choose which playground to open the code in, using the
-  playground name.
+  the user will get to choose which playground to open the code in.
 * The `Language` field is the human-readable Pygments language name for that
   programming language. The syntax highlighting language in a code block
   is internally mapped to these human-readable Pygments names.
   E.g: `py3` and `py` are mapped to `Python`. We are working on implementing a
-  typeahead for looking up the Pygments name. Till then, one can use
-  [this Pygments method](https://pygments-doc.readthedocs.io/en/latest/lexers/lexers.html#pygments.lexers.get_lexer_by_name) as a workaround.
-* We also support this feature for languages outside of Pygments. If you have
-  a custom language, you can simply hook up a custom playground to it and specify
-  it as the code block syntax highlighter.
+  typeahead for looking up the Pygments name. Until then, one can use
+  [this Pygments method](https://pygments-doc.readthedocs.io/en/latest/lexers/lexers.html#pygments.lexers.get_lexer_by_name).
+* The links for opening code playgrounds are always constructed by
+  concatenating the provided URL prefix with the URL-encoded contents
+  of the code block.
+* Code playground sites do not always clearly document their URL
+  format; often you can just get the prefix from your browser's URL bar.
 
 If you have any trouble setting in setting up a code playground, please [contact
 us](/help/contact-support) with details on what you're trying to do, and we'll be
diff --git a/templates/zerver/help/format-your-message-using-markdown.md b/templates/zerver/help/format-your-message-using-markdown.md
index 10003ea424..23e0306539 100644
--- a/templates/zerver/help/format-your-message-using-markdown.md
+++ b/templates/zerver/help/format-your-message-using-markdown.md
@@ -118,8 +118,9 @@ Organization administrators can also configure a default syntax
 highlighting language.  In this configuration, one can use ````text`
 to display content without any syntax highlighting.
 
-See [adding a playground option](/help/add-a-custom-playground) to open this Markdown
-code in external playgrounds based on its syntax highlighting language.
+Hovering over a Zulip code block reveals buttons to copy the code or
+open it in one of the organization's configured [custom code
+playgrounds](/help/add-a-custom-playground).
 
 ## LaTeX
 ~~~
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 3f49c1092b..3c1a5fbc76 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -2629,8 +2629,8 @@ paths:
                               additionalProperties: false
                               description: |
                                 Event sent to all users in a Zulip organization when the
-                                set of configured [playgrounds](/help/add-a-custom-playground) for
-                                the organization has changed.
+                                set of configured [code playgrounds](/help/add-a-custom-playground)
+                                for the organization has changed.
 
                                 **Changes**: New in Zulip 4.0 (feature level 49).
                               properties:
@@ -6986,9 +6986,7 @@ paths:
       operationId: add_realm_playground
       tags: ["server_and_organizations"]
       description: |
-        Configure realm [playgrounds](/help/add-a-custom-playground) options
-        to run code snippets occuring in a code block using playgrounds which
-        supports that language.
+        Configure [code playgrounds](/help/add-a-custom-playground) for the organization.
 
         `POST {{ api_url }}/v1/realm/playgrounds`
 
@@ -7044,8 +7042,8 @@ paths:
       operationId: remove_realm_playground
       tags: ["server_and_organizations"]
       description: |
-        Remove realm [playgrounds](/help/add-a-custom-playground) options to
-        run code snippets in custom playgrounds.
+        Remove a [code playground](/help/add-a-custom-playground) previously
+        configured for an organization.
 
         `DELETE {{ api_url }}/v1/realm/playgrounds/{playground_id}`
 
@@ -7451,7 +7449,8 @@ paths:
                         description: |
                           Present if `realm_playgrounds` is present in `fetch_event_types`.
 
-                          An array of dictionaries where each dictionary describes a [playgrounds](/help/add-a-custom-playground) entry in this Zulip organization.
+                          An array of dictionaries where each dictionary describes a
+                          [code playground](/help/add-a-custom-playground) configured for this Zulip organization.
 
                           **Changes**: New in Zulip 4.0 (feature level 49).
                       realm_user_groups:
```

I think it's worth your thinking about each of those changes, since most of them are things that would be worth incorporating into your technical documentation communication style.

Thanks @sumanthvrao!

sumanthvrao commented:
Yaay 🎉 ! Thanks for the documentation improvements, I'll keep them in mind going forward! :)
